File: apex/position.py
"""Exit management — SL/TP, breakeven, trailing, runner mode.

Pure functions over a position dict — used identically by the live loop
(apex/bot.py) and the backtest (backtest.py), so what you backtest is what
trades.
"""
from apex import config as cfg
from apex import forex, logger


def calc_sltp(side, price, atr_value, symbol):
    if cfg.ATR_BASED_SL and atr_value > 0:
        sl_dist, tp_dist = atr_value * cfg.ATR_SL_MULT, atr_value * cfg.ATR_TP_MULT
    else:
        sl_dist = forex.from_pips(cfg.STOP_LOSS_PIPS, symbol)
        tp_dist = forex.from_pips(cfg.TAKE_PROFIT_PIPS, symbol)
    return {"stopLoss": price - sl_dist if side == "BUY" else price + sl_dist,
            "takeProfit": price + tp_dist if side == "BUY" else price - tp_dist}


def check_position(pos, price):
    """Mutates pos in-place (trailing, breakeven, runner) and returns the close
    trigger ('STOP_LOSS' | 'TAKE_PROFIT' | 'TRAIL_PROFIT') or None."""
    if not pos:
        return None
    side = pos["side"]
    symbol = pos.get("symbol", cfg.SYMBOL)
    entry = pos["entryPrice"]

    # Breakeven (PTJ): la +1R mută SL la entry ±1 pip — trade-ul nu mai poate pierde
    if cfg.BREAKEVEN_AT_R > 0 and not pos.get("beDone") and pos.get("initialStop"):
        one_r = abs(entry - pos["initialStop"]) * cfg.BREAKEVEN_AT_R
        pip = forex.from_pips(1, symbol)
        be_price = entry + pip if side == "BUY" else entry - pip
        if ((side == "BUY" and price >= entry + one_r and be_price > pos["stopLoss"])
                or (side == "SELL" and price <= entry - one_r and be_price < pos["stopLoss"])):
            pos["stopLoss"] = be_price
            pos["beDone"] = True
            logger.info(f"🛡️ Breakeven: SL moved to {be_price:.5f} — risk-free trade")

    # Trailing stop (strâns în runner mode — Seykota: let profits run)
    if cfg.TRAILING_STOP:
        trail_pips = cfg.RUNNER_TRAIL_PIPS if pos.get("runner") else cfg.TRAILING_STOP_PIPS
        trail_dist = forex.from_pips(trail_pips, symbol)
        if side == "BUY":
            pos["trailHigh"] = max(pos.get("trailHigh") or price, price)
            pos["stopLoss"] = max(pos["stopLoss"], pos["trailHigh"] - trail_dist)
        else:
            pos["trailLow"] = min(pos.get("trailLow") or price, price)
            pos["stopLoss"] = min(pos["stopLoss"], pos["trailLow"] + trail_dist)

    pnl_pips = forex.to_pips(price - entry if side == "BUY" else entry - price, symbol)
    pos["pnlPips"] = round(pnl_pips, 1)

    hit_sl = price <= pos["stopLoss"] if side == "BUY" else price >= pos["stopLoss"]
    hit_tp = price >= pos["takeProfit"] if side == "BUY" else price <= pos["takeProfit"]
    if hit_sl:
        return "TRAIL_PROFIT" if pos.get("runner") else "STOP_LOSS"
    if hit_tp and not pos.get("runner"):
        # Runner doar pe paper — la live brokerul execută TP-ul server-side oricum
        if cfg.LET_WINNERS_RUN and cfg.TRAILING_STOP and cfg.PAPER_TRADING:
            pos["runner"] = True
            logger.info(f"🏃 TP hit at {price} — runner mode: letting profit run "
                        f"(trail {cfg.RUNNER_TRAIL_PIPS:g} pips)")
            return None
        return "TAKE_PROFIT"
    return None

File: apex-trade-bot/src/position.js
/**
 * Exit management — SL/TP, breakeven, trailing, runner mode.
 */
const cfg    = require('./config');
const logger = require('./logger');

function calcSLTP(side, price, atrValue) {
  if (cfg.ATR_BASED_SL && atrValue > 0) {
    const slDist = atrValue * cfg.ATR_SL_MULT;
    const tpDist = atrValue * cfg.ATR_TP_MULT;
    return {
      stopLoss:   side === 'BUY' ? price - slDist : price + slDist,
      takeProfit: side === 'BUY' ? price + tpDist : price - tpDist,
    };
  }
  return {
    stopLoss:   side === 'BUY' ? price * (1 - cfg.STOP_LOSS_PCT)   : price * (1 + cfg.STOP_LOSS_PCT),
    takeProfit: side === 'BUY' ? price * (1 + cfg.TAKE_PROFIT_PCT) : price * (1 - cfg.TAKE_PROFIT_PCT),
  };
}

function checkPosition(pos, price) {
  if (!pos) return null;
  const { side, entryPrice, takeProfit } = pos;

  if (cfg.BREAKEVEN_AT_R > 0 && !pos.beDone && pos.initialStop) {
    const oneR = Math.abs(entryPrice - pos.initialStop) * cfg.BREAKEVEN_AT_R;
    const bePrice = side === 'BUY' ? entryPrice * (1 + 2 * cfg.FEE_PCT) : entryPrice * (1 - 2 * cfg.FEE_PCT);
    if ((side === 'BUY' && price >= entryPrice + oneR && bePrice > pos.stopLoss) ||
        (side === 'SELL' && price <= entryPrice - oneR && bePrice < pos.stopLoss)) {
      Object.assign(pos, { stopLoss: bePrice, beDone: true });
      logger.info(`🛡️ Breakeven: SL moved to $${bePrice.toFixed(5)} — risk-free trade`);
    }
  }

  if (cfg.TRAILING_STOP) {
    const dist = pos.runner ? cfg.RUNNER_TRAIL_DIST : cfg.TRAILING_STOP_DIST;
    if (side === 'BUY') {
      pos.trailHigh = Math.max(pos.trailHigh ?? price, price);
      pos.stopLoss  = Math.max(pos.stopLoss, pos.trailHigh * (1 - dist));
    } else {
      pos.trailLow  = Math.min(pos.trailLow ?? price, price);
      pos.stopLoss  = Math.min(pos.stopLoss, pos.trailLow * (1 + dist));
    }
  }

  pos.pnlPct = (side === 'BUY' ? price - entryPrice : entryPrice - price) / entryPrice * 100;

  const hitSL = side === 'BUY' ? price <= pos.stopLoss : price >= pos.stopLoss;
  const hitTP = side === 'BUY' ? price >= takeProfit   : price <= takeProfit;

  if (hitSL) return pos.runner ? 'TRAIL_PROFIT' : 'STOP_LOSS';
  if (hitTP && !pos.runner) {
    if (cfg.LET_WINNERS_RUN && cfg.TRAILING_STOP && cfg.PAPER_TRADING) {
      pos.runner = true;
      logger.info(`🏃 TP reached at $${price} — runner mode: letting profits run (trail ${cfg.RUNNER_TRAIL_DIST * 100}%)`);
      return null;
    }
    return 'TAKE_PROFIT';
  }
  return null;
}

module.exports = { calcSLTP, checkPosition };
